modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/TreeIterator.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.persistence.defragmentation;
+
+import java.nio.ByteBuffer;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.internal.processors.cache.persistence.pagemem.PageMemoryEx;
+import org.apache.ignite.internal.processors.cache.persistence.tree.BPlusTree;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.BPlusIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.BPlusLeafIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.BPlusMetaIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.PageIO;
+import org.apache.ignite.internal.util.GridUnsafe;
+
+/** */
+public class TreeIterator {
+    /** Direct memory buffer with a size of one page. */
+    private final ByteBuffer pageBuf;
+
+    /** Offheap page size. */
+    private final int pageSize;
+
+    /** */
+    public TreeIterator(int size) {
+        pageSize = size;
+
+        pageBuf = ByteBuffer.allocateDirect(pageSize);
+    }
+
+    // Performance impact of constant closures allocation is not clear. So this method should be avoided in massive
+    // operations like tree leaves access.
+    /** */
+    public static <T> T access(
+        PageAccessType access,
+        PageMemoryEx pageMemory,
+        int grpId,
+        long pageId,
+        PageAccessor<T> accessor
+    ) throws IgniteCheckedException {
+        assert access != null;
```

Comment:
Should we add assert here `PageAccessType.ACCESS_READ | PageAccessType.ACCESS_WRITE`? In case of extending `PageAccessType` we will get an incorrect results

Reply:
I doubt that anyone will add anything, these types correspond to "lockRead" and "lockWrite" for the pages.
Technically there is a "restore write lock", or whatever it is called, so I'll change the assert, thank you!